### ISSUE TITLE
Docker compose version 2 file format breaks the http proxy

### ISSUE THREAD

**Issue by marclennox** [codekitchen/dinghy#135]
I tried switching to docker compose v2 file format, which has the services indented underneath a `services` key.  It appears that this breaks Dinghy's ability to expose virtual hosts using the http proxy container.

**Comment by codekitchen** [comment]
I haven't had a chance to look into this yet but I'm guessing it's related to nginx-proxy needing an upgrade (https://github.com/codekitchen/dinghy/issues/120)

**Comment by marclennox** [comment]
Seems strange though because the environment variables are being passed through just the same 

Sent from Outlook Mobile




On Tue, Feb 16, 2016 at 12:38 PM -0800, "Brian Palmer" <notifications@github.com> wrote:












I haven't had a chance to look into this yet but I'm guessing it's related to nginx-proxy needing an upgrade (#120)

**Comment by codekitchen** [comment]
The problem isn't with the env vars, it's because using a v2 compose file triggers compose to use docker's new networking layer, which nginx-proxy doesn't support yet.

https://github.com/jwilder/nginx-proxy/issues/362

Looks like there's a couple open PRs to fix this but none have been merged in yet. I'm not planning to work around this in Dinghy, I'd rather wait for the fix in nginx-proxy. Hopefully it doesn't take long.

**Comment by marclennox** [comment]
Ah right, this makes sense.  I remember reading something about docker no longer exposing environment variables in the new networking scheme, which a lot of containers rely on.

**Comment by marclennox** [comment]
Agreed though, no point fixing in dinghy!

**Comment by codekitchen** [comment]
Now that I've had some time to dig into the problem, I see that this will take some changes in dinghy even once nginx-proxy is fixed. The problem is we want our proxy to automatically work with any container that sets VIRTUAL_HOST, and docker-compose v2 now creates a private network for each app. So our proxy has to configure itself to join those private networks as necessary, so that it can reach the containers.

I think I have an idea on how to make this happen by writing a wrapper `-notify` script, but I doubt it'll be accepted into the upstream nginx-proxy project since it's not really their intended use case. I'll take a whack at it soon.

**Comment by codekitchen** [comment]
Support for this has landed on dinghy's master branch. I want to let it bake for a bit before releasing it, I need to test it out in more varied scenarios. You can try it out by installing the [dinghy prerelease](https://github.com/codekitchen/dinghy#prereleases), you should see the proxy using `codekitchen/dinghy-http-proxy:2.0.0` (version 2.0.0) in `docker ps`.

**Comment by marclennox** [comment]
Great stuff.  I just quickly tested it out.

It appears as though the proper configuration are being placed into the `/etc/nginx/conf.d/default.conf` however when I browse to any of the sites, I get the "Welcome to nginx on Debian!" page.

**Comment by codekitchen** [comment]
Hm interesting, that's the default "I don't know about the domain you gave me" page, which would imply the domain isn't in the config. Should be a `server_name <mydomain>.docker` block in the config. If you've confirmed that's in the config, maybe check for any errors in `/var/log/nginx/{error,access}.log` in the container?

**Comment by marclennox** [comment]
error.log is showing errors when nginx tried to reload the updated configuration.

`nginx: [emerg] no host in upstream ":9292" in /etc/nginx/conf.d/default.conf:28`

The line in question is in the upstream definition, and there's no host

```
upstream justreply.docker {
        server :9292;
}
```

**Comment by marclennox** [comment]
Note that all of the upstream definitions are without a host.

**Comment by codekitchen** [comment]
Ah that's definitely the problem thanks. I haven't seen that locally though, the logic is to grab the first exported address or network from each container, see https://github.com/codekitchen/dinghy-http-proxy/blob/master/nginx.tmpl#L46

Would it be possible to reduce your docker-compose.yml down to something you'd be comfortable pasting here? I bet a solid repro would make it easy to track down.

**Comment by codekitchen** [comment]
Never mind on that @marclennox , I found a repro in one of my local compose files. I'll fix it up.

**Comment by marclennox** [comment]
For sure, here's a pared down version that exhibits the issue.

```
version: '2'

volumes:
  postgresql_data:
    external: false
  rails_bundle:
    external: false

services:

  #
  # Volume container for all persisted data
  #
  volumes:
    image: tianon/true:latest
    volumes:
      - ..:/usr/src/app
      - postgresql_data:/data/postgresql
      - rails_bundle:/data/bundle

  #
  # Key-value database used for ephemeral storage and backs the sidekiq queues.
  # This is a required service to run the stack
  #
  redis_server:
    hostname: redis-server
    image: tenstartups/alpine-redis:latest

  #
  # Caching service not normally required in development unless specifically testing
  # centralized caching.
  #
  memcache_server:
    hostname: memcache-server
    image: tenstartups/alpine-memcached:latest

  #
  # Main database service that is required service to run the stack
  #
  postgresql_server:
    hostname: postresql-server
    environment:
      PGDATA: /data/postgresql
    image: postgres:latest
    volumes_from:
      - volumes

  #
  # Rails web server service that serves up all dynamic content and is required
  # to run the stack.
  #
  web_server:
    hostname: web-server
    environment:
      - VIRTUAL_HOST=justreply.docker
      - VIRTUAL_PORT=9292
      - RAILS_ENV
      - PRY_LISTEN_ADDRESS=web-server
      - PRY_LISTEN_PORT=9876
      - BUNDLE_APP_CONFIG=.bundle
      - BUNDLE_PATH=/data/bundle
      - ENV_FILE=config/application.env
    image: tenstartups/rails:latest
    command: web-server
    volumes_from:
      - volumes
    ports:
      - 9292 # Puma HTTP port
    links:
      - redis_server
      - memcache_server
      - postgresql_server
```

**Comment by codekitchen** [comment]
Thanks! An update to dinghy so that it uses `codekitchen/dinghy-http-proxy:2.0.1` will fix it, I switched that function call from `or` to `coalesce` because that described my intent better, not realizing that `coalesce` treats the empty string differently than `or` does in this templating language.

**Comment by marclennox** [comment]
Works!  I'll put it through it's paces on a few of my projects tomorrow.

Thanks @codekitchen

**Comment by marclennox** [comment]
Found another issue.  It seems that virtual hosts are being created even though I don't set the environment variable, based on the inferred name of the service.  In some cases, this is causing a duplicate upstream to be created, which causes the nginx config to fail.

So for example, I have a service called `web_server` which sets the environment variable `VIRTUAL_HOST=justreply.docker`

I end up with 2 upstreams (`web_server.justreply.docker` and `justreply.docker`)

That doesn't cause failure, but I have other service called `ngrok` which sets the environment variable `VIRTUAL_HOST=ngrok.justreply.docker`

In this case I get 2 identical upstreams (`ngrok.justreply.docker`) which causes it to fail.

I'm assuming the "inferred" upstreams should not be showing up?

**Comment by codekitchen** [comment]
Ah yeah, that's actually intentional, I'm experimenting with the idea of proxying to any compose container via `<service>.<project>.docker` so that defining a `VIRTUAL_HOST` isn't necessary for local dev. I didn't document it because I haven't decided whether to keep it yet.

I hadn't tested the case where it overlaps with an explicit VIRTUAL_HOST, that makes sense that it'd break. For an immediate workaround you can remove the VIRTUAL_HOST definition, I'll look into a fix.

**Comment by marclennox** [comment]
OK cool.  I worked around it fine by removing the VIRTUAL_HOST environment.  I wonder if the behaviour should be that you don't create the "default" proxy if VIRTUAL_HOST is explicitly provided.

**Comment by codekitchen** [comment]
Yep, I just pushed up v2.0.2 of the proxy that does just that. Thanks again for testing.

**Comment by marclennox** [comment]
OK, I no longer get duplicate upstreams if I specify a VIRTUAL_HOST that is the same name as the service, however I do get 2 (different) upstreams if a specify a VIRTUAL_HOST that is different from the service name.  Is that you intended behaviour?

**Comment by codekitchen** [comment]
Yeah I went back and forth on that, but decided there was probably no harm in just allowing access via both hostnames.

**Comment by marclennox** [comment]
Seems fine to me, just wanted to make sure.  I'll keep testing, but it looks like it's working well for my projects now.

**Comment by michaellopez** [comment]
I've had success with HEAD version too. Simple tests have all worked for our use cases.

**Comment by codekitchen** [comment]
I'm going to close this issue as the new proxy has been released in version 4.3. Thanks again for all the help!
